projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+func main() {
+	c := http.Client{Timeout: time.Duration(1) * time.Second}
+
+	resp, err := c.Get("http://localhost:8080")
+
+	if err != nil {
+		fmt.Fprint(os.Stderr, "Something went wrong. Cannot complete request at the moment\n")
+		return
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case 429:
+		retryTime := resp.Header.Get("Retry-After")
+		formattedRetryTime, err := parseRetryValue(retryTime)
+
+		if err != nil {
+			// cannot determine how long to sleep for, retry once after 1 sec delay. Reason - The server response shows it is safe to retry. However we might not want keep at it for long - hence the choice
+
+			sleep(1)
+			resp, err = retry(1, c)
+
+			if err != nil {
+				// retry still unsuccessful and server does not respond
+				fmt.Fprint(os.Stderr, "Something went wrong. Cannot complete request at the moment\n")
+				resp.Body.Close()
+				return
+			}
+
+		} else if formattedRetryTime > 5 {
+			// sleep time is long
+			fmt.Fprint(os.Stderr, "Server busy. Cannot retrieve weather\n")
+			resp.Body.Close()
+			return
+		} else {
+			sleep(formattedRetryTime)
+			resp, err = retry(3, c) // chose 3 as max number of retries
+
+			if err != nil {
+				// retry still unsuccessful and server does not respond
+				fmt.Fprint(os.Stderr, "Something went wrong. Cannot complete request at the moment\n")
+				resp.Body.Close()
+				return
+			}
+
+		}
+	}
+
+	// for other responses - print to stdout
+	body, _ := io.ReadAll(resp.Body)
+	fmt.Fprint(os.Stdout, string(body), "\n")
+}
+
+func parseRetryValue(v string) (int, error) {
+
+	value, err := strconv.Atoi(v)
+	if err != nil {
+		// check if retry time is in httpTime format
+		httpTime, httpTimeErr := time.Parse(time.RFC1123, v)
+
+		if httpTimeErr == nil {
+			value = int(time.Until(httpTime).Seconds())
+			err = httpTimeErr
+		}
+	}
+	return value, err
+
+}
+
+func retry(maxRetries int, c http.Client) (*http.Response, error) {
```

Comment:
There's quite a bit of shared code between this function and the `main` function - how you make a request, the idea of retrying on failure, determining success of a request/response - as well as some duplicated code (both functions do some sleeping between retries).

I'd recommend trying to structure your code so that you just have one loop, you just have one place in the loop where you make HTTP requests, and you just have one place in the loop where you handle responses (and maybe sleep). This gives you greater control (e.g. it lets you be really specific about how long you're sleeping between retries, rather than having two different bits of code each contributing a bit), and means your code will act consistently (e.g. your retries don't currently parse the `Retry-After` header - but they probably should!)

Give some thought as to what's special about retries vs the first attempt (e.g. you print "Retrying ...") and work out how to make that code only happen for retries, while trying to structure your code to have one source of truth for what "attempting to get the weather" looks like.

Reply:
@illicitonion I have given some thought about this and restructured my code. I believe it also reads better than the previous one as just looking at the main function, I could easily see and understand what is going on.